### PR TITLE
Fix LDAP injection bug

### LINKED ISSUE #2356: LDAP injection attacks are possible through user field
LDAP password authenticator doesn't seem to escape special LDAP characters, which makes it possible to change the logic of LDAP searches and binds.

See [Testing for LDAP Injection](https://www.owasp.org/index.php/Testing_for_LDAP_Injection_(OTG-INPVAL-006)) or [Understanding and Exploiting Web-based LDAP](https://pen-testing.sans.org/blog/2017/11/27/understanding-and-exploiting-web-based-ldap) for examples and [LDAP Injection Prevention Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/LDAP_Injection_Prevention_Cheat_Sheet.html) for possible fixes.

### PR DESCRIPTION
**Pull request by MiguelWeezardo**
Fixes #2356 by encoding special LDAP characters

### CODE REVIEW

File: presto-password-authenticators/src/test/java/io/prestosql/plugin/password/ldap/TestLdapEncoding.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestLdapEncoding
+{
+    @Test
+    void testDNEncoding()
+    {
+        // DNs which don't require encoding
+        assertThat(LdapAuthenticator.encodeForDN("I can eat glass and it doesn't hurt me"))
```
Comment (electrum) [inline presto-password-authenticators/src/test/java/io/prestosql/plugin/password/ldap/TestLdapEncoding.java]: This text is a bit strange (some may find it offensive). Let’s use something more generic.

File: presto-password-authenticators/src/main/java/io/prestosql/plugin/password/ldap/LdapAuthenticator.java [hunk 2]
```diff
@@ -171,9 +173,88 @@ private void checkForGroupMembership(String user, DirContext context)
...
+        }
+        return sb.toString();
+    }
+
+    // Based on https://www.owasp.org/index.php/Preventing_LDAP_Injection_in_Java
+    @VisibleForTesting
+    static String encodeForDN(String input)
+    {
+        StringBuilder sb = new StringBuilder();
+        if ((input.length() > 0) && ((input.charAt(0) == ' ') || (input.charAt(0) == '#'))) {
```
Comment (electrum) [inline presto-password-authenticators/src/main/java/io/prestosql/plugin/password/ldap/LdapAuthenticator.java]: Use `isEmpty`

File: presto-password-authenticators/src/main/java/io/prestosql/plugin/password/ldap/LdapAuthenticator.java [hunk 3]
```diff
@@ -171,9 +173,88 @@ private void checkForGroupMembership(String user, DirContext context)
...
+            sb.append('\\'); // add the leading backslash if needed
+        }
+        for (int i = 0; i < input.length(); i++) {
+            char c = input.charAt(i);
+            switch (c) {
+                case '\\':
```
Comment (electrum) [inline presto-password-authenticators/src/main/java/io/prestosql/plugin/password/ldap/LdapAuthenticator.java]: These can simplify to
```java
case ...:
case ...:
  sb.append(‘\\’).append(c);

File: presto-password-authenticators/src/main/java/io/prestosql/plugin/password/ldap/LdapAuthenticator.java (hunk 2 above)
Comment (MiguelWeezardo) [reply presto-password-authenticators/src/main/java/io/prestosql/plugin/password/ldap/LdapAuthenticator.java]: I just found an existing method javax.naming.ldap.Rdn.escapeValue() which works for DN attributes. I didn't find anything similar for escaping attributes in search queries.

File: presto-password-authenticators/src/main/java/io/prestosql/plugin/password/ldap/LdapAuthenticator.java (hunk 3 above)
Comment (MiguelWeezardo) [reply presto-password-authenticators/src/main/java/io/prestosql/plugin/password/ldap/LdapAuthenticator.java]: I just found an existing method javax.naming.ldap.Rdn.escapeValue() which works for DN attributes. I didn't find anything similar for escaping attributes in search queries.

**Comment by MiguelWeezardo** [comment]
@electrum If you think this version is acceptable, please merge it into master. I don't have merge permissions yet.

File: presto-password-authenticators/src/main/java/io/prestosql/plugin/password/ldap/LdapAuthenticator.java [hunk 4]
```diff
@@ -171,9 +174,46 @@ private void checkForGroupMembership(String user, DirContext context)
         }
     }
 
-    private static String replaceUser(String pattern, String user)
+    @VisibleForTesting
+    static String replaceUserInSearch(String pattern, String user)
     {
-        return pattern.replaceAll("\\$\\{USER}", user);
+        return pattern.replaceAll("\\$\\{USER}", Matcher.quoteReplacement(encodeForSearch(user)));
```
Comment (findepi) [inline presto-password-authenticators/src/main/java/io/prestosql/plugin/password/ldap/LdapAuthenticator.java]: Use `String.replace`. You don't need regular expressions.

File: presto-password-authenticators/src/main/java/io/prestosql/plugin/password/ldap/LdapAuthenticator.java [hunk 5]
```diff
@@ -171,9 +174,46 @@ private void checkForGroupMembership(String user, DirContext context)
...
+    }
+
+    @VisibleForTesting
+    static String replaceUserInBind(String pattern, String user)
+    {
+        return pattern.replaceAll("\\$\\{USER}", Matcher.quoteReplacement(Rdn.escapeValue(user)));
```
Comment (findepi) [inline presto-password-authenticators/src/main/java/io/prestosql/plugin/password/ldap/LdapAuthenticator.java]: Use `String.replace`. You don't need regular expressions.

File: presto-password-authenticators/src/test/java/io/prestosql/plugin/password/ldap/TestLdapEncoding.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+package io.prestosql.plugin.password.ldap;
+
+import org.testng.annotations.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestLdapEncoding
+{
+    @Test
+    void testSearchEncoding()
```
Comment (findepi) [inline presto-password-authenticators/src/test/java/io/prestosql/plugin/password/ldap/TestLdapEncoding.java]: `public`, `testEncodeForSearch`

File: presto-password-authenticators/src/test/java/io/prestosql/plugin/password/ldap/TestLdapEncoding.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.password.ldap;
+
+import org.testng.annotations.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestLdapEncoding
```
Comment (findepi) [inline presto-password-authenticators/src/test/java/io/prestosql/plugin/password/ldap/TestLdapEncoding.java]: while current class name is reasonable, it's not easily find-able.
change to `TestLdapAuthenticator`

File: presto-password-authenticators/src/test/java/io/prestosql/plugin/password/ldap/TestLdapEncoding.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+{
+    @Test
+    void testSearchEncoding()
+    {
+        // Search queries which don't require encoding - pangrams from http://clagnut.com/blog/2380
```
Comment (findepi) [inline presto-password-authenticators/src/test/java/io/prestosql/plugin/password/ldap/TestLdapEncoding.java]: remove comment; is not necessary to understand the code

File: presto-password-authenticators/src/test/java/io/prestosql/plugin/password/ldap/TestLdapEncoding.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        assertThat(LdapAuthenticator.encodeForSearch("*"))
+                .as("Wildcard encoded")
+                .isEqualTo("\\2a");
+        assertThat(LdapAuthenticator.encodeForSearch("Hi (This) = is * a \\ test # ç à ô"))
+                .as("Multiple special characters encoded")
+                .isEqualTo("Hi \\28This\\29 = is \\2a a \\5c test # ç à ô");
+    }
+
+    @Test
+    void testDNInjections()
```
Comment (findepi) [inline presto-password-authenticators/src/test/java/io/prestosql/plugin/password/ldap/TestLdapEncoding.java]: `public`, `testReplaceUserInBind`

**Comment by findepi** [comment]
Did you consider owasp advice
> The best way to prevent LDAP injection is to use a positive validation scheme for ensuring that the data going into your queries doesn't contain any attacks.

?

**Comment by findepi** [comment]
Discussed offline. Since previously user names with "special characters" were not allowed (i.e. they would exploit or malform search query), we will explicitly outlaw them now, instead of escaping.
I am concerned that even proper escaping in our code may expose some downstream problems/bugs, for no user-visible value.
We will reconsider escaping when explicit need arises.

File: presto-password-authenticators/src/main/java/io/prestosql/plugin/password/ldap/LdapAuthenticator.java [hunk 10]
```diff
@@ -55,6 +56,7 @@
         implements PasswordAuthenticator
 {
     private static final Logger log = Logger.get(LdapAuthenticator.class);
+    private static final String SPECIAL_LDAP_CHARACTERS = ",=+<>#;*()\"\\\u0000";
```
Comment (findepi) [inline presto-password-authenticators/src/main/java/io/prestosql/plugin/password/ldap/LdapAuthenticator.java]: Use `com.google.common.base.CharMatcher#anyOf`

Then use `CharMatcher#matchesAnyOf` for checking if rogue char is present

File: presto-password-authenticators/src/main/java/io/prestosql/plugin/password/ldap/LdapAuthenticator.java [hunk 11]
```diff
@@ -126,6 +131,31 @@ private Principal authenticate(String user, String password)
...
+    static boolean containsSpecialLDAPCharacters(String user)
+    {
+        if (!user.isEmpty()) {
+            // whitespace at beginning and end is considered special
+            char c = user.charAt(0);
+            if (c == ' ' || c == '\r') {
+                return true;
+            }
+            c = user.charAt(user.length() - 1);
+            if (c == ' ' || c == '\r') {
```
Comment (findepi) [inline presto-password-authenticators/src/main/java/io/prestosql/plugin/password/ldap/LdapAuthenticator.java]: Use CharMatcher for this as well

File: presto-password-authenticators/src/main/java/io/prestosql/plugin/password/ldap/LdapAuthenticator.java [hunk 12]
```diff
@@ -126,6 +131,31 @@ private Principal authenticate(String user, String password)
         }
     }
 
+    // Based on https://www.owasp.org/index.php/Preventing_LDAP_Injection_in_Java and javax.naming.ldap.Rdn.escapeStringValue
+    @VisibleForTesting
+    static boolean containsSpecialLDAPCharacters(String user)
```
Comment (findepi) [inline presto-password-authenticators/src/main/java/io/prestosql/plugin/password/ldap/LdapAuthenticator.java]: remove `LDAP` from method name. We're in scope of LDAP-specific class. 

add javadoc like
```
 /**
  * Returns {@code true} when parameter contains a character that has a special meaning in
  * LDAP search or bind name (DN).
 */
```

File: presto-password-authenticators/src/main/java/io/prestosql/plugin/password/ldap/LdapAuthenticator.java [hunk 13]
```diff
@@ -173,7 +203,7 @@ private void checkForGroupMembership(String user, DirContext context)
 
     private static String replaceUser(String pattern, String user)
     {
-        return pattern.replaceAll("\\$\\{USER}", user);
+        return pattern.replace("${USER}", user);
```
Comment (findepi) [inline presto-password-authenticators/src/main/java/io/prestosql/plugin/password/ldap/LdapAuthenticator.java]: move to separate commit

**Pull request merged**